File: flowcatalyst-go/internal/platform/common/result.go
// Package common provides core infrastructure for the UseCase/UnitOfWork pattern.
// This package enforces that all successful mutations go through UnitOfWork.Commit(),
// guaranteeing that domain events and audit logs are always emitted.
package common

// Result represents the outcome of a use case execution.
// Success can ONLY be created via UnitOfWork.Commit() to guarantee
// domain events and audit logs are always emitted.
//
// The success constructor (newSuccess) is unexported, meaning only code within
// this package (specifically MongoUnitOfWork) can create successful results.
// This is the key enforcement mechanism that guarantees every successful
// mutation emits a domain event and creates an audit log.
type Result[T any] struct {
	value   T
	err     *UseCaseError
	success bool
}

// newSuccess creates a successful result.
// This is unexported - only UnitOfWork.Commit() can create successful results,
// ensuring domain events are always emitted.
func newSuccess[T any](value T) Result[T] {
	return Result[T]{
		value:   value,
		success: true,
	}
}

// Failure creates a failed result.
// This is public - any code can return failures for validation errors,
// business rule violations, etc. without going through UnitOfWork.
func Failure[T any](err *UseCaseError) Result[T] {
	return Result[T]{
		err:     err,
		success: false,
	}
}

// FailureFrom creates a failed result from an existing error.
// Convenience method for wrapping errors.
func FailureFrom[T any](err *UseCaseError) Result[T] {
	return Failure[T](err)
}

// IsSuccess returns true if the result is successful.
func (r Result[T]) IsSuccess() bool {
	return r.success
}

// IsFailure returns true if the result is a failure.
func (r Result[T]) IsFailure() bool {
	return !r.success
}

// Value returns the success value.
// Should only be called after checking IsSuccess().
func (r Result[T]) Value() T {
	return r.value
}

// Error returns the error if the result is a failure, nil otherwise.
func (r Result[T]) Error() *UseCaseError {
	return r.err
}

// Map transforms a successful result's value using the provided function.
// If the result is a failure, it returns the failure unchanged.
func Map[T, U any](r Result[T], fn func(T) U) Result[U] {
	if r.IsFailure() {
		return Failure[U](r.err)
	}
	return newSuccess(fn(r.value))
}

// FlatMap chains result-returning operations.
// If the result is a failure, it returns the failure unchanged.
// If successful, it applies the function which returns a new Result.
func FlatMap[T, U any](r Result[T], fn func(T) Result[U]) Result[U] {
	if r.IsFailure() {
		return Failure[U](r.err)
	}
	return fn(r.value)
}

// Match applies one of two functions depending on success/failure state.
// This is useful for handling both cases in a single expression.
func Match[T, U any](r Result[T], onSuccess func(T) U, onFailure func(*UseCaseError) U) U {
	if r.IsSuccess() {
		return onSuccess(r.value)
	}
	return onFailure(r.err)
}

// OrElse returns the success value or the provided default if failure.
func (r Result[T]) OrElse(defaultValue T) T {
	if r.IsSuccess() {
		return r.value
	}
	return defaultValue
}

// OrElseGet returns the success value or calls the provided function if failure.
func (r Result[T]) OrElseGet(fn func() T) T {
	if r.IsSuccess() {
		return r.value
	}
	return fn()
}

File: flowcatalyst-typescript/packages/domain-core/src/result.ts
/**
 * Result Type for Use Case Execution
 *
 * This is a discriminated union with two variants:
 * - Success<T> - contains the successful result value
 * - Failure<T> - contains the error details
 *
 * **IMPORTANT:** The `success()` factory is restricted. Only UnitOfWork can create
 * successful results, guaranteeing that domain events are always emitted when
 * state changes.
 *
 * Usage in use cases:
 * ```typescript
 * // Return failure for validation/business rule violations
 * if (!isValid) {
 *     return Result.failure(UseCaseError.validation('INVALID', 'Invalid input'));
 * }
 *
 * // Return success only through UnitOfWork.commit()
 * return unitOfWork.commit(aggregate, event, command);
 * ```
 *
 * Usage in API layer:
 * ```typescript
 * if (result.isSuccess()) {
 *     return c.json(result.value, 200);
 * } else {
 *     return c.json(result.error, UseCaseError.httpStatus(result.error));
 * }
 * ```
 */

import type { UseCaseError } from './errors.js';

/**
 * Token for authorizing Result.success() creation.
 * Only UnitOfWork implementations should have access to this token.
 *
 * @internal
 */
export const RESULT_SUCCESS_TOKEN: unique symbol = Symbol('RESULT_SUCCESS_TOKEN');

/**
 * Type for the success authorization token.
 *
 * @internal
 */
export type ResultSuccessToken = typeof RESULT_SUCCESS_TOKEN;

/**
 * Successful result containing the value.
 */
export interface Success<T> {
	readonly _tag: 'success';
	readonly value: T;
}

/**
 * Failed result containing the error.
 */
export interface Failure<T> {
	readonly _tag: 'failure';
	readonly error: UseCaseError;
}

/**
 * Result type - either Success or Failure.
 */
export type Result<T> = Success<T> | Failure<T>;

/**
 * Type guard to check if a result is a success.
 */
export function isSuccess<T>(result: Result<T>): result is Success<T> {
	return result._tag === 'success';
}

/**
 * Type guard to check if a result is a failure.
 */
export function isFailure<T>(result: Result<T>): result is Failure<T> {
	return result._tag === 'failure';
}

/**
 * Result factory functions.
 */
export const Result = {
	/**
	 * Create a successful result.
	 *
	 * **RESTRICTED:** Requires the success token. Only UnitOfWork should call this.
	 * Use cases must return success through `unitOfWork.commit()`.
	 *
	 * @param token - The success authorization token (only available to UnitOfWork)
	 * @param value - The success value
	 * @returns A Success result
	 * @throws Error if token is invalid
	 *
	 * @internal
	 */
	success<T>(token: ResultSuccessToken, value: T): Success<T> {
		if (token !== RESULT_SUCCESS_TOKEN) {
			throw new Error(
				'Result.success() is restricted. Use UnitOfWork.commit() to create successful results. ' +
					'This ensures domain events and audit logs are always created with state changes.',
			);
		}
		return { _tag: 'success', value } as Success<T>;
	},

	/**
	 * Create a failed result.
	 *
	 * This is public - any code can create failures for validation errors,
	 * business rule violations, etc.
	 *
	 * @param error - The use case error
	 * @returns A Failure result
	 */
	failure<T>(error: UseCaseError): Failure<T> {
		return { _tag: 'failure', error } as Failure<T>;
	},

	/**
	 * Check if a result is a success.
	 */
	isSuccess,

	/**
	 * Check if a result is a failure.
	 */
	isFailure,

	/**
	 * Map a successful result to a new value.
	 *
	 * @param result - The result to map
	 * @param fn - The mapping function
	 * @returns A new result with the mapped value
	 */
	map<T, U>(result: Result<T>, fn: (value: T) => U): Result<U> {
		if (isSuccess(result)) {
			// Note: We create a new Success directly here since we're just transforming
			// an existing success, not creating one from a use case
			return { _tag: 'success', value: fn(result.value) } as Success<U>;
		}
		return { _tag: 'failure', error: result.error } as Failure<U>;
	},

	/**
	 * Match on a result, handling both success and failure cases.
	 *
	 * @param result - The result to match
	 * @param onSuccess - Handler for success case
	 * @param onFailure - Handler for failure case
	 * @returns The result of the matched handler
	 */
	match<T, U>(result: Result<T>, onSuccess: (value: T) => U, onFailure: (error: UseCaseError) => U): U {
		if (isSuccess(result)) {
			return onSuccess(result.value);
		}
		return onFailure(result.error);
	},

	/**
	 * Get the value from a success result, or throw an error.
	 *
	 * @param result - The result to unwrap
	 * @returns The success value
	 * @throws Error if the result is a failure
	 */
	unwrap<T>(result: Result<T>): T {
		if (isSuccess(result)) {
			return result.value;
		}
		throw new Error(`Cannot unwrap failure result: ${result.error.code} - ${result.error.message}`);
	},

	/**
	 * Get the value from a success result, or return a default.
	 *
	 * @param result - The result to unwrap
	 * @param defaultValue - The default value if result is a failure
	 * @returns The success value or the default
	 */
	unwrapOr<T>(result: Result<T>, defaultValue: T): T {
		if (isSuccess(result)) {
			return result.value;
		}
		return defaultValue;
	},
};
